### ISSUE TITLE
torchrun: Hostname/endpoint mismatch not handled

### ISSUE THREAD

**Issue by janEbert** [pytorch/pytorch#73656]
### 🐛 Describe the bug

TL;DR: I need to monkey patch `torchrun` to be able to handle a `rendezvous_endpoint` with a format different from `socket.gehostname()`.

# Description

I have a very peculiar setup for a Slurm cluster: if I want to communicate across InfiniBand cells, the addresses must have an "i" appended. Let's assume node adresses follow the regular expression format `node[0-9]` (or `node[0-9]i`). Optionally, the name of the cluster may follow these addresses, so we get `node[0-9].cluster` or `node[0-9]i.cluster`.

## Correct Case
I'm starting jobs on nodes 0 to 3, with `node0` being the master node; so I pass `--rendezvous_endpoint node0` to `torchrun`. Everything works fine because all nodes are on the same InfiniBand cell and don't require the "i" appended.

## Failure Case
I now start a job on nodes 0 to 7. Assume 0 to 3 are on one InfiniBand cell, 4 to 7 on another; I thus have to append the "i" so the nodes can communicate between one another. I pass `--rendezvous_endpoint node0i` to `torchrun`.

### First Error
With the `"c10d"` backend, PyTorch [tries to check whether we are host](https://github.com/pytorch/pytorch/blob/v1.10.2/torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py#L144) using `sockets.gethostname()` and some other variants.
In the [`_matches_machine_hostname` function](https://github.com/pytorch/pytorch/blob/v1.10.2/torch/distributed/elastic/rendezvous/utils.py#L121-L139), these variables get the following values:
- `addr` becomes `None` due to an error (can't parse an IP address from our endpoint hostname),
- `this_host` becomes `"node0"` or similar,
- the single `addr_info` in `addr_list` contains a hostname similar to "node0.cluster" at index 3.

Now, the host argument to that function is `"node0i"` (note the suffixed "i") as I want to communicate over InfiniBand cells.
This means all the comparisons fail for all nodes, meaning PyTorch doesn't find a host! PyTorch expects a host to be found, so this is where we run into a deadlock and consequent timeout.

#### First Solution
However, luckily there is an out [via some extra config values for rendezvous backends](https://github.com/pytorch/pytorch/blob/v1.10.2/torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py#L136-L144). The section I linked is the original `if`-query from which the (in my case erroneous) `_matches_machine_hostname` is called.
If I specify an `"is_host"` config, I avoid that call; `torchrun` has the `--rdzv_conf` argument for these config values, so it's just a matter of passing it in correctly depending on if `torchrun` is started from the master node or any other.

### Second Error
Now the script starts but I run into another issue when `torch.distributed.init_process_group` is called. That's because [the `MASTER_ADDR` envvar is set](https://github.com/pytorch/pytorch/blob/v1.10.2/torch/distributed/elastic/agent/server/api.py#L514-L515) based [on `socket.gethostname()` again](https://github.com/pytorch/pytorch/blob/v1.10.2/torch/distributed/elastic/agent/server/api.py#L384-L385) (the end result looks like `addr_info[3]` from before, so `_get_fq_hostname() == 'node0.cluster`)! Note the missing "i" suffix – the nodes can't communicate across InfiniBand cells again, so I run into another deadlock.

Now you could argue that I can avoid this erroneous `_get_fq_hostname` result by passing a `--master_addr` to `torchrun` so that it's not `None` in the `if`-query. However, [that argument is ignored unless the backend is `"static"`](https://github.com/pytorch/pytorch/blob/v1.10.2/torch/distributed/launcher/api.py#L169-L173); the `MASTER_ADDR` will be obtained dynamically in any other case. So as a non-root user I have two options:

#### Second Solution
1. Monkey patch `_get_fq_hostname` to append an "i" in the `torchrun` script. Advantage is that I can use `torchrun` for any other scripts as well without any further changes.
2. Modify my script and insert the "i" in the `MASTER_ADDR` envvar at the correct location. Advantage is that I don't have to patch `torchrun` for each different Python environment.

## Possible Fixes
Having an "out" for the second error case like another `--rdzv_conf` flag or not ignoring `--master_addr` would be nice. Currently, solutions are involved and very manual. Another option would be to allow methods other than `socket.gethostname` for grabbing the host addresses (in both `_matches_machine_hostname` and `_get_fq_hostname`).

### Versions

1.10.2

**Comment by kiukchung** [comment]
can you standardize the address format with the `i` suffix? That is, in the same cell, could nodes still reach each other if you used `node[0-9]i.cluster`?

**Comment by janEbert** [comment]
Hey, thanks for a possible solution! This does indeed work. but only as long as all nodes are inside one InfiniBand cell. Once a client node is on another InfiniBand cell, it cannot communicate via NCCL to the server on `node0.cluster` and connections time out.

**Comment by kiukchung** [comment]
@janEbert Two follow up questions:

1) with my proposal above, do the nodes timeout with NCCL or during torchelastic's rendezvous?
2) the nodes timeout even when addressing node0 as `node0i.cluster`?

**Comment by janEbert** [comment]
I could have been wrong about the communication over specifically NCCL not working, sorry.

1. The timeout happens in `init_process_group` [when the `TCPStore` is created](https://github.com/pytorch/pytorch/blob/v1.11.0/torch/distributed/rendezvous.py#L160-L162).
2. I'll get back to you once my test job has been queued. Just to be clear, you mean the following (adding the "i" to master node as well)?
    On master node: `torchrun --rdzv_backend=c10d --rdzv_endpoint=node0i.cluster:29500 --rdzv_conf=is_host`
    Everywhere else: `torchrun --rdzv_backend=c10d --rdzv_endpoint=node0i.cluster:29500`

**Comment by kiukchung** [comment]
ah I see the issue you are having with `--rdzv_backend=c10d`, its "auto assigning" the `MASTER_ADDR` by getting the fully-qualified hostname on node rank 0. This is b/c there are two "masters" in torchrun:

1. the rendezvous master (this is a the node level) which is used to coordinate the agents on each node. This can be assigned by the user with `rendezvous_endpoint` (and need not actually be one of the nodes - e.g. it could be an etcd endpoint with the `etcd` rdzv backend)
2. the `MASTER_ADDR` (rank master) that is passed to the worker and is used by `dist.init_process_group()`. This currently cannot be set by the user. This is be design since torchelastic was originally built to handle node elasticity, hence the "rank" master has to be assigned by the elastic agent since rank 0's host could be scaled down and replaced with another node.

the `static` rendezvous is special (since it was added for backwards compatibility with `torch.distributed.launch`) and so is `c10d` (since it uses one of the nodes as the rendezvous endpoint - it is essentially master fault-intolerant). 

There's two fixes:

1. Have you tried using `--rdzv_backend=static` and providing 
     1. `--master_addr=node0.cluster` on the nodes in the same cell as `node0`
     2. `--master_addr=node0i.cluster` on the nodes in different cells
     3. `--rank=$SLURM_ENV_VAR_FOR_TASK_INDEX` (I believe slurm has an env var that represents the index of the task in the job).
     4. `--world_size=$SLURM_ENV_VAR_FOR_TOTAL_NUM_TASKS` (you might have to multiply `SLURM_NNODES * SLURM_NTASKS_PER_NODE`)

2. In https://github.com/pytorch/pytorch/blob/cda8635a5e9822f08440f4ef89f7e70a6977b1fe/torch/distributed/launcher/api.py#L160 add another if-statement to check `rdzv_backend=="c10d"` as follows:

```
    endpoint = rdzv_parameters.endpoint
    endpoint = endpoint.strip()
    if not endpoint:
         raise ValueError(
             "Endpoint is missing in endpoint. Specify `--rdzv_endpoint=$host:$port`
         )
    master_addr, master_port = parse_rendezvous_endpoint(endpoint, default_port=-1)

    if rdzv_parameters.backend == "c10d" and rdzv_parameters.get_as_bool("is_host"):
        # use the same master as rendezvous master for rank master (MASTER_ADDR passed to the worker script)
        # port has to be None so that a free random port is assigned for MASTER_PORT
        return (master_addr, None) 
    elif rdzv_parameters.backend != "static":
        return (None, None)
    else: # backend == "static"       
        if master_port == -1:
            raise ValueError(
                f"port is missing in endpoint: {endpoint}. Try to specify --master_port"
            )
        return (master_addr, master_port)
 ```

**Comment by janEbert** [comment]
You summarized the "two masters" problem really well, thanks!

I think for your first suggestion, the same problem applies as when I did:

> On master node: torchrun --rdzv_backend=c10d --rdzv_endpoint=node0.cluster:29500 --rdzv_conf=is_host
> Everywhere else: torchrun --rdzv_backend=c10d --rdzv_endpoint=node0i.cluster:29500

The problem being that the server with non-"i'-suffixed name cannot be seen from different InfiniBand cells.

The second solution seems super promising, though. I'll try that out and get back to you.

**Comment by janEbert** [comment]
I'll probably not be able to inform you until next week. One question I have is, is it important that non-`is_host` nodes have `master_addr is None`? Or in other words, why should non-host nodes not use the same `master_addr` parsed from the endpoint?

**Comment by janEbert** [comment]
The second error can recently be avoided by specifying `--local-addr=<master-addr>` (with `<master-addr>` being the same as specified in `--rdzv_endpoint <master-addr>:<port>`). This means monkey patching is no longer necessary!

Once again, here is an updated complete solution for a Slurm cluster on more recent PyTorch versions:

1. Add `--rdzv_conf=is_host=$(if ((SLURM_NODEID)); then echo 0; else echo 1; fi)` to the `torchrun` arguments. It's important that the value of `SLURM_NODEID` is populated from _inside_ the `srun` command. Wrapping the whole `torchrun ...` execution like `bash -c 'torchrun ...'` may help for this.
2. Add `--local_addr=<master_addr>`, where `<master_addr>` is the same as specified in `--rdzv_endpoint=<master-addr>:<port>`.

**Comment by avolkov1** [comment]
> ah I see the issue you are having with `--rdzv_backend=c10d`, its "auto assigning" the `MASTER_ADDR` by getting the fully-qualified hostname on node rank 0. This is b/c there are two "masters" in torchrun:
> 
> 1. the rendezvous master (this is a the node level) which is used to coordinate the agents on each node. This can be assigned by the user with `rendezvous_endpoint` (and need not actually be one of the nodes - e.g. it could be an etcd endpoint with the `etcd` rdzv backend)
> 2. the `MASTER_ADDR` (rank master) that is passed to the worker and is used by `dist.init_process_group()`. This currently cannot be set by the user. This is be design since torchelastic was originally built to handle node elasticity, hence the "rank" master has to be assigned by the elastic agent since rank 0's host could be scaled down and replaced with another node.
> 
> the `static` rendezvous is special (since it was added for backwards compatibility with `torch.distributed.launch`) and so is `c10d` (since it uses one of the nodes as the rendezvous endpoint - it is essentially master fault-intolerant).
> 
> There's two fixes:
> 
> 1. Have you tried using `--rdzv_backend=static` and providing
>    
>    1. `--master_addr=node0.cluster` on the nodes in the same cell as `node0`
>    2. `--master_addr=node0i.cluster` on the nodes in different cells
>    3. `--rank=$SLURM_ENV_VAR_FOR_TASK_INDEX` (I believe slurm has an env var that represents the index of the task in the job).
>    4. `--world_size=$SLURM_ENV_VAR_FOR_TOTAL_NUM_TASKS` (you might have to multiply `SLURM_NNODES * SLURM_NTASKS_PER_NODE`)
> 2. In https://github.com/pytorch/pytorch/blob/cda8635a5e9822f08440f4ef89f7e70a6977b1fe/torch/distributed/launcher/api.py#L160
>     add another if-statement to check `rdzv_backend=="c10d"` as follows:
> 
> ```
>     endpoint = rdzv_parameters.endpoint
>     endpoint = endpoint.strip()
>     if not endpoint:
>          raise ValueError(
>              "Endpoint is missing in endpoint. Specify `--rdzv_endpoint=$host:$port`
>          )
>     master_addr, master_port = parse_rendezvous_endpoint(endpoint, default_port=-1)
> 
>     if rdzv_parameters.backend == "c10d" and rdzv_parameters.get_as_bool("is_host"):
>         # use the same master as rendezvous master for rank master (MASTER_ADDR passed to the worker script)
>         # port has to be None so that a free random port is assigned for MASTER_PORT
>         return (master_addr, None) 
>     elif rdzv_parameters.backend != "static":
>         return (None, None)
>     else: # backend == "static"       
>         if master_port == -1:
>             raise ValueError(
>                 f"port is missing in endpoint: {endpoint}. Try to specify --master_port"
>             )
>         return (master_addr, master_port)
> ```

This should be up streamed.

**Comment by geomlyd** [comment]
> The second error can recently be avoided by specifying `--local-addr=<master-addr>` (with `<master-addr>` being the same as specified in `--rdzv_endpoint <master-addr>:<port>`). This means monkey patching is no longer necessary!
> 
> Once again, here is an updated complete solution for a Slurm cluster on more recent PyTorch versions:
> 
>     1. Add `--rdzv_conf=is_host=$(if ((SLURM_PROCID)); then echo 0; else echo 1; fi)` to the `torchrun` arguments. It's important that the value of `SLURM_PROCID` is populated from _inside_ the `srun` command. Wrapping the whole `torchrun ...` execution like `bash -c 'torchrun ...'` may help for this.
> 
>     2. Add `--local_addr=<master_addr>`, where `<master_addr>` is the same as specified in `--rdzv_endpoint=<master-addr>:<port>`.

Hey there, I'm facing similar problems as the ones you describe above. I'm running PyTorch 2.1.2 and while I do use your suggestions for 1 (btw, it should explicitly be SLURM_PROCID and not SLURM_NODEID, right?) and 2 it seems that my processes still won't communicate over Infiniband. Do I still need to modify the actual PyTorch source code and/or do something else besides using torchrun with (1) and (2)?

**Comment by janEbert** [comment]
Hey! I'm pretty sure in previous versions `SLURM_NODEID` was correct and that also matches comments in the code regarding the `is_host` option referring to the current _node_, but with the recent versions I have to use `SLURM_PROCID` as you said; otherwise the initial rendezvous cannot be established. I did not face any errors with `SLURM_PROCID` so it seems like the correct choice recently. I also liberally edited the comment you're citing to include that change, which did say `SLURM_NODEID` before a few days ago. :)

For me, it was enough to use 1. and 2., which I love because we can avoid monkey patching. What environment are you in, and are you also fighting with a different hostname being required for InfiniBand communication?
One potential issue you may be facing, is if `$SLURM_PROCID` is expanded too early. However, in that case you'd get a timeout for the initial rendezvous.

Feel free to send me the Slurm submission script you're using, I can take a look whether I spot anything.

**Comment by janEbert** [comment]
Great to hear that the fix had worked for you!
One thing that you did not explicitly include but which is also important is that you have `--ntasks-per-node=1`, because in our case we delegate starting the per-node parallel processes to `torchrun` instead of `srun`.

Other than that everything else looks correct to me, so I'm curious why it would not work. :/

**Comment by janEbert** [comment]
Hey @geomlyd, I can reproduce. There are indeterministic hangs for me upon the second rendezvous. The issue happens in the `TCPStore` code, which is in the C++ part of the code base. This fact already makes debugging much more annoying for me, so I hope that there's something obvious hidden in the code. :/

**Comment by geomlyd** [comment]
Hey, thanks for the update on this. So in short this means that your suggested code fixes are still necessary in PyTorch 2, and should be used in tandem with the new arguments?

**Comment by janEbert** [comment]
Basically it's still the same solution as for previous PyTorch versions, yes. For minimal monkey patching, you'd use `--rdzv_conf=is_host=$(if ((SLURM_NODEID)); then echo 0; else echo 1; fi)` and fix up the 2nd issue from the initial description here. I don't have tested code ready for that at the moment, but it can be figured out from the branch with the fix: https://github.com/janEbert/pytorch/commits/fix-hostname-endpoint-mismatch/

**Comment by janEbert** [comment]
Maybe just this solution still works, even for PyTorch 2: https://github.com/pytorch/pytorch/issues/73656#issuecomment-1131699879

**Comment by janEbert** [comment]
FYI I wrote the [`torchrun_jsc`](https://github.com/HelmholtzAI-FZJ/torchrun_jsc) package to fix this issue in a simple way for users, since there is resistance to merging the solution for some reason. It can be `pip`-installed and usage is described in the README (it's really just replacing `torchrun` with `torchrun_jsc`).

**Comment by janEbert** [comment]
Good news everyone! Starting with PyTorch 2.5, the following seems not to cause race conditions, meaning monkey patching is finally _actually_ no longer necessary:

You need to add the following arguments on the host node and only on the host node:  
`--rdzv_conf=is_host=1 --local_addr=<master_addr>`, where `<master_addr>` is the same as specified in `--rdzv_endpoint=<master-addr>:<port>`.

For a solution for Slurm similar to the previous ones:

1. Add `--rdzv_conf=is_host="$(if ((SLURM_NODEID)); then echo 0; else echo 1; fi)"` to the `torchrun` arguments. It's important that the value of `SLURM_NODEID` is populated from _inside_ the `srun` command. Wrapping the whole `torchrun ...` execution like `bash -c 'torchrun ...'` may help for this.
2. Add `--local_addr="$(if ((SLURM_NODEID)); then echo <master_addr>; else hostname; fi)"`, where `<master_addr>` is the same as specified in `--rdzv_endpoint=<master-addr>:<port>`. Similar to before, `SLURM_NODEID` (and in this case, `hostname` as well) needs to be populated from inside the `srun` command.

`torchrun_jsc` now defaults to this on supported versions as the less intrusive fix instead of monkey patching.

**Comment by K-H-Ismail** [comment]
The solution presented here works https://github.com/pytorch/pytorch/issues/73656#issuecomment-2669740505 in practice for a single node and multi-gpus, but I still encounter this error when using at least 2 nodes.

```bash
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtitan/lib/python3.10/site-packages/torch/distributed/run.py", line 909, in run
    elastic_launch(
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtitan/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 138, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtitan/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 260, in launch_agent
    result = agent.run()
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtitan/lib/python3.10/site-packages/torch/distributed/elastic/metrics/api.py", line 137, in wrapper
    result = f(*args, **kwargs)
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtitan/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 711, in run
    result = self._invoke_run(role)
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtitan/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 864, in _invoke_run
    self._initialize_workers(self._worker_group)
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtitan/lib/python3.10/site-packages/torch/distributed/elastic/metrics/api.py", line 137, in wrapper
    result = f(*args, **kwargs)
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtitan/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 683, in _initialize_workers
    self._rendezvous(worker_group)
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtitan/lib/python3.10/site-packages/torch/distributed/elastic/metrics/api.py", line 137, in wrapper
    result = f(*args, **kwargs)
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtitan/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 500, in _rendezvous
    rdzv_info = spec.rdzv_handler.next_rendezvous()
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtitan/lib/python3.10/site-packages/torch/distributed/elastic/rendezvous/dynamic_rendezvous.py", line 1206, in next_rendezvous
    self._bootstrap_store_info = RendezvousStoreInfo.build(
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtitan/lib/python3.10/site-packages/torch/distributed/elastic/rendezvous/api.py", line 95, in build
    addr = store.get(RendezvousStoreInfo.MASTER_ADDR_KEY).decode(encoding="UTF-8")
torch.distributed.DistStoreError: wait timeout after 60000ms, keys: /torch.rendezvous.11468275.0/MASTER_ADDR
```

**Comment by janEbert** [comment]
Is this with PyTorch 2.7? Could be that APIs were modified again, breaking the fix.

**Comment by K-H-Ismail** [comment]
2.7 definitely, 2.6.0 as well. With the patch [torchrun_jsc](https://github.com/HelmholtzAI-FZJ/torchrun_jsc) there is no problem. Only torchrun with the aforementioned --rdzv_conf=is_host=1 and --local_addr=<master_addr>. Here is the full stack trace using an example from the torchtune library.

```bash
`pyenv activate' requires Pyenv and Pyenv-Virtualenv to be loaded into your shell.
Check your shell configuration and Pyenv and Pyenv-Virtualenv installation instructions.

ESC[0mW0508 19:42:29.877000 440008 /my_path/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/run.py:792] 
W0508 19:42:29.877000 440008 /my_path/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/run.py:792] *****************************
************
W0508 19:42:29.877000 440008 /my_path/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/run.py:792] Setting OMP_NUM_THREADS envir
onment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
W0508 19:42:29.877000 440008 /my_path/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/run.py:792] *****************************
************
[W508 19:42:29.081090059 socket.cpp:759] [c10d] The client socket cannot be initialized to connect to [jwb0129i.juwels]:21068 (errno: 97 - Address family not supported by protocol).
W0508 19:42:42.007000 2769425 /my_path/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/run.py:792] 
W0508 19:42:42.007000 2769425 /my_path/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/run.py:792] ****************************
*************
W0508 19:42:42.007000 2769425 /my_path/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/run.py:792] Setting OMP_NUM_THREADS envi
ronment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
W0508 19:42:42.007000 2769425 /my_path/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/run.py:792] ****************************
*************
[W508 19:42:42.473919643 socket.cpp:759] [c10d] The client socket cannot be initialized to connect to [jwb0129i.juwels]:21068 (errno: 97 - Address family not supported by protocol).
[W508 19:42:47.546752777 socket.cpp:759] [c10d] The client socket cannot be initialized to connect to [jwb0129.juwels]:33937 (errno: 97 - Address family not supported by protocol).
[W508 19:43:47.342623299 socket.cpp:464] [c10d] waitForInput: poll for socket SocketImpl(fd=58, addr=jwb0129i.juwels:40882, remote=jwb0129i.juwels:21068) returned 0, likely a timeout
[W508 19:43:47.342680839 socket.cpp:489] [c10d] waitForInput: socket SocketImpl(fd=58, addr=jwb0129i.juwels:40882, remote=jwb0129i.juwels:21068) timed out after 60000ms

Traceback (most recent call last):
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/torchtune/bin/tune", line 8, in <module>
    sys.exit(main())
  File "/my_path/official/torchtune/torchtune/_cli/tune.py", line 52, in main
    parser.run(args)
  File "/my_path/official/torchtune/torchtune/_cli/tune.py", line 46, in run
    args.func(args)
  File "/my_path/official/torchtune/torchtune/_cli/run.py", line 212, in _run_cmd
    self._run_distributed(args, is_builtin=is_builtin)
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 355, in wrapper
    return f(*args, **kwargs)
  File "/my_path/official/torchtune/torchtune/_cli/run.py", line 101, in _run_distributed
    run(args)
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/run.py", line 909, in run
    elastic_launch(
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 138, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 260, in launch_agent
    result = agent.run()
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/elastic/metrics/api.py", line 137, in wrapper
    result = f(*args, **kwargs)
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 711, in run
    result = self._invoke_run(role)
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 864, in _invoke_run
    self._initialize_workers(self._worker_group)
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/elastic/metrics/api.py", line 137, in wrapper
    result = f(*args, **kwargs)
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 683, in _initialize_workers
    self._rendezvous(worker_group)
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/elastic/metrics/api.py", line 137, in wrapper
    result = f(*args, **kwargs)
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/elastic/agent/server/api.py", line 500, in _rendezvous
    rdzv_info = spec.rdzv_handler.next_rendezvous()
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/elastic/rendezvous/dynamic_rendezvous.py", line 1206, in next_rendezvous
    self._bootstrap_store_info = RendezvousStoreInfo.build(
  File "/p/home/jusers/khalfaoui1/juwels/.pyenv/versions/3.10.13/envs/torchtune/lib/python3.10/site-packages/torch/distributed/elastic/rendezvous/api.py", line 95, in build
    addr = store.get(RendezvousStoreInfo.MASTER_ADDR_KEY).decode(encoding="UTF-8")
torch.distributed.DistStoreError: wait timeout after 60000ms, keys: /torch.rendezvous.11468275.0/MASTER_ADDR
srun: error: jwb0149: task 1: Terminated
srun: error: jwb0129: task 0: Exited with exit code 1
srun: Force Terminated StepId=11468275.0
```
the slurm script was

```bash
#!/usr/bin/env bash

#SBATCH --output=slurm_%j.out
#SBATCH --error=slurm_%j.err
#SBATCH --nodes=2
#SBATCH --ntasks-per-node=1
# 96 CPUs w/o nomultithread
# 48 CPUs w/ nomultithread
#SBATCH --cpus-per-task=24
# Use only physical cores.
#SBATCH --hint=nomultithread
#SBATCH --gres=gpu:4
#SBATCH --time=00:05:00

curr_file="$(scontrol show job "$SLURM_JOB_ID" | grep '^[[:space:]]*Command=' | head -n 1 | cut -d '=' -f 2-)"
curr_dir="$(dirname "$curr_file")"


# Propagate the specified number of CPUs per task to each `srun`.
export SRUN_CPUS_PER_TASK="$SLURM_CPUS_PER_TASK"

#GCCcore/.12.3.0
pyenv activate 3.10.13/torchtune

if [ "$SYSTEMNAME" = juwelsbooster ] \
       || [ "$SYSTEMNAME" = juwels ] \
       || [ "$SYSTEMNAME" = jurecadc ] \
       || [ "$SYSTEMNAME" = jusuf ]; then
    # Allow communication over InfiniBand cells on JSC machines.
    MASTER_ADDR="$MASTER_ADDR"i
fi
export MASTER_PORT=21068

# Prevent NCCL not figuring out how to initialize.
export NCCL_SOCKET_IFNAME=ib0
# Prevent GLOO not being able to communicate.
export GLOO_SOCKET_IFNAME=ib0


# For the example, we get a bunch of information from the
# checkpointing function about tensors being deduplicated. With this
# line, we turn off logging for those warnings. Please do not keep
# this for your scripts if you do not encounter the same problem.
# For more information, see:
# https://github.com/pytorch/pytorch/issues/117392
export TORCH_LOGS='-torch.distributed.checkpoint._dedup_tensors'
export HF_DATASETS_OFFLINE=1
export HF_HUB_OFFLINE=1


export NCCL_IB_TIMEOUT=50
export UCX_RC_TIMEOUT=4s
export NCCL_IB_RETRY_CNT=10
export NCCL_P2P_DISABLE=1
export CUDA_LAUNCH_BLOCKING=1

srun env -u CUDA_VISIBLE_DEVICES tune run \
       --nproc_per_node=gpu \
       --nnodes="$SLURM_JOB_NUM_NODES" \
       --rdzv_id="$SLURM_JOB_ID" \
       --rdzv_endpoint="$MASTER_ADDR":"$MASTER_PORT" \
       --rdzv_backend=c10d \
       --rdzv_conf=is_host="$(if ((SLURM_NODEID)); then echo 0; else echo 1; fi)" \
       --local_addr="$(if ((SLURM_NODEID)); then echo $MASTER_ADDR; else hostname; fi)" \
       full_finetune_distributed \
       --config ./recipes/configs/llama3_1/8B_full_custom.yaml \
```

**Comment by janEbert** [comment]
All right, I see the problem; this is exactly why the "manual" solution is tricky. :)

Highlighting what you're missing from the solution:

1. Add `--rdzv_conf=is_host="$(if ((SLURM_NODEID)); then echo 0; else echo 1; fi)"` to the `torchrun` arguments. **It's important that the value of `SLURM_NODEID` is populated from _inside_ the `srun` command. Wrapping the whole `torchrun ...` execution like `bash -c 'torchrun ...'` may help for this.**
2. Add `--local_addr="$(if ((SLURM_NODEID)); then echo <master_addr>; else hostname; fi)"`, where `<master_addr>` is the same as specified in `--rdzv_endpoint=<master-addr>:<port>`. **Similar to before, `SLURM_NODEID` (and in this case, `hostname` as well) needs to be populated from inside the `srun` command.**

Right now, the variables are being populated from _outside_ `srun`; they are evaluated on the root node that executes the `sbatch` script, but _not_, as they should be, on each individual node.
What should work is the following (just wrapping in `bash -c '...'`:

```shell
srun env -u CUDA_VISIBLE_DEVICES bash -c 'tune run \
       --nproc_per_node=gpu \
       --nnodes="$SLURM_JOB_NUM_NODES" \
       --rdzv_id="$SLURM_JOB_ID" \
       --rdzv_endpoint="$MASTER_ADDR":"$MASTER_PORT" \
       --rdzv_backend=c10d \
       --rdzv_conf=is_host="$(if ((SLURM_NODEID)); then echo 0; else echo 1; fi)" \
       --local_addr="$(if ((SLURM_NODEID)); then echo $MASTER_ADDR; else hostname; fi)" \
       full_finetune_distributed \
       --config ./recipes/configs/llama3_1/8B_full_custom.yaml'
```

I noticed your script doesn't seem to be setting `MASTER_ADDR` before the "i" change. `MASTER_ADDR` would additionally need to be `export`ed for the above to work.

**Comment by K-H-Ismail** [comment]
Confirmed, wrapping in  ```bash -c ' ... ' ``` and setting ```--rdzv_conf``` and ```--local_addr``` as described above fixed it, thanks!

**Comment by matbun** [comment]
@janEbert , to complement on your last comment, on JSC Juwels booster partition I recently tested the following options: 

1. `--rdzv_conf=is_host=$(( SLURM_NODEID == 0 ? 1 : 0 ))`
2. `--local_addr="$(hostname -s)i"`

And they seem to work (up to 50 nodes, which is more than an IB cell) with torch 2.6 and `torchrun_jsc` launcher.

**Comment by janEbert** [comment]
Thanks @matbun! Just for information and so you can clean up your scripts: you don't need to add these settings when using the `torchrun_jsc` launcher, because it adds them automatically. That also means that – unless it's required in another place – you can remove a `bash -c '...'` wrapping. :)
